### ISSUE TITLE
fix(training): report real token deltas and make optimizer reports discoverable

### ISSUE THREAD

**Issue by lalalune** [elizaOS/eliza#14876]
## Problem

Nothing renders optimization results beyond the raw `vN.json` artifact and logs. GEPA computes a full (score, prompt_token_count) Pareto frontier and then discards everything but the best-score member; promotion-gate decisions (incumbent stddev, threshold, verdict) are invisible; there is no way to see WHAT changed in a promoted prompt, which provider earned its tokens, or which trajectories drove a win. Operators cannot audit or trust the auto-training loop they are asked to leave enabled.

## Current state (refs)

- Artifacts: `<stateDir>/optimized-prompts/<task>/vN.json` with `lineage[{round, variant, score, notes}]` and fewShotExamples already persisted (`packages/core/src/services/optimized-prompt.ts:158`); run records at `<state>/training/runs/<runId>.json` (`plugins/plugin-training/src/core/training-orchestrator.ts`); rejected candidates land as `candidate_rejected_<ts>.json` (`promotion-gate.ts`).
- Frontier: `plugins/plugin-training/src/optimizers/gepa.ts` returns only the best-score frontier member; the rest is discarded.
- Existing builders to extend, not duplicate: `plugins/plugin-training/src/core/eval-comparison-artifact.ts`, `benchmark-matrix-artifact.ts`, `training-readiness-report.ts`; scenario-runner run viewer; `packages/scripts/trajectory.ts` (has stage prompt byte/segment diff, no artifact-diff mode).
- Surface: Settings Auto-training panel (`packages/ui/src/components/settings/CapabilitiesSection.tsx`, `/api/training/auto/config|status`) currently shows status only.

## Proposed report format (one self-contained HTML/MD artifact per run)

(a) Per-task baseline->optimized prompt diff + lineage timeline (data already in the artifact); (b) provider-inclusion ablation table — per provider: delta judge_score, delta prompt tokens, verdict (requires the trajectory-attribution companion issue); (c) rendering-variant comparison — same seeded data, N render templates, scored side by side; (d) quality-vs-tokens Pareto chart — persist the FULL frontier from gepa.ts into the artifact (e.g. `artifact.frontier[]`) instead of best-only, chart promoted point vs frontier; (e) promotion-gate decision trace — incumbent reseed scores, mean, stddev, 1.5-sigma threshold, verdict, and the rejected-candidate pointer when applicable.

## Where it surfaces

1. Written next to the run record: `<state>/training/runs/<runId>.html` (+ machine-readable `report.json` section inside the existing run record) — same pattern as scenario-runner reports.
2. Linked from the Settings Auto-training panel per run (extend `/api/training/auto/status` to carry the report path/URL); panel renders verdict + headline deltas inline.
3. `packages/scripts/trajectory.ts artifact-diff <task> [--from vN --to vM]` for terminal use; report doubles as the PR_EVIDENCE attachment for optimizer PRs.

## Acceptance criteria + evidence bar

- `bun run train -- --backend native --task <t> --optimizer gepa` produces the HTML+JSON report; all five sections render with real data from a live run (attached to the PR and read by hand).
- gepa.ts persists the full frontier; promotion + rejection paths both produce decision traces; panel links resolve.
- Report is self-contained (no external fetches) and renders in light/dark.

## Out of scope

The attribution capture itself (companion issue); new optimizer algorithms; cloud dashboards — local state-dir + Settings panel only.

**Comment by lalalune** [comment]
Project 15 claim: I am taking this from Ready to In progress. I will scope the first pass to deterministic report data/rendering that can be validated locally, and will leave the issue In progress if live GEPA/model evidence cannot be captured in this environment.

**Comment by lalalune** [comment]
Project 15 status update: opened PR #15100 for deterministic optimization run reports.

Implemented in the PR:
- GEPA now returns its final quality-vs-token Pareto frontier, and promoted optimized-prompt artifacts preserve that frontier.
- Promotion-gate decisions are persisted on promoted artifacts; rejected candidates still write `candidate_rejected_<ts>.json` with gate scores.
- Auto-training run records now get sibling `report.json` and `report.html` files under `<state>/training/runs/<runId>/`, plus report paths/headline fields in the top-level run JSON.
- `/api/training/auto/status` now includes recent report paths/summaries for the settings panel to link/render.
- Report sections for provider ablations and rendering variants are explicit unavailable states when the required attribution/seeded-render data is absent.

Local evidence reviewed:
- `bun run --cwd plugins/plugin-training test -- src/optimizers/__tests__/gepa.test.ts src/core/optimization-report.test.ts src/core/promotion-persist.test.ts src/core/training-orchestrator.test.ts src/routes/training-routes.test.ts` -> 5 files / 21 tests passed
- `bun run --cwd packages/core test -- src/services/optimized-prompt.test.ts` -> 1 file / 30 tests passed
- `bun x biome check ...` on touched files
- `bun run --cwd plugins/plugin-training build`
- `bun run --cwd packages/core build:node`
- `bun run --cwd packages/core typecheck`
- `git diff --check`
- `bun run audit:error-policy-ratchet`
- `node scripts/check-pr-evidence.mjs --body-file /tmp/pr-14876.md`

Known gap: I am leaving this issue In progress because the acceptance bar still needs a real `bun run train -- --backend native --task <task> --optimizer gepa` run with live model credentials, generated report artifacts, logs, and manual review. `plugins/plugin-training typecheck` still fails on the package's ambient Node/workspace declaration and pre-existing strictness diagnostics; the focused tests and package build covering the touched paths pass.

**Comment by lalalune** [comment]
MVP/project evidence-state refresh (2026-07-09): implementation PR #15100 is merged and the issue already records deterministic validation for report JSON/HTML generation, GEPA frontier persistence, promotion traces, and status API report links.

Remaining acceptance is a real `bun run train -- --backend native --task <task> --optimizer gepa` run with live model/provider credentials, generated `report.json`/`report.html`, logs, and manual review of all report sections.

Current shell has no live model/cloud credentials, so I cannot produce that training run evidence here. Added `needs-shaw`; keep open until credentialed report artifacts and the hand-read summary are attached.

**Comment by lalalune** [comment]
Project triage correction: moved from **Needs human review** to **Ready**. Reporting scaffolding still marks core dimensions unavailable; real task-linked prompt diffs, provider ablations, Pareto data, trajectory attribution, and rendered report proof remain engineering work.

**Comment by lalalune** [comment]
Triage disposition: narrow this to reporting correctness and discoverability; the broad report catalog is not an MVP workstream. The generated report's headline `tokenDelta` is currently `optimizedPrompt.length - baselinePrompt.length`—a character delta—while the HTML labels it “Token/char delta.” `providerAblations` and `renderingVariants` are permanently emitted as unavailable, so their presence does not satisfy the original acceptance criteria. The original Settings-link and trajectory artifact-diff contracts also were not delivered by the merged implementation.

Acceptance should be: (1) report an actual token metric from attributable usage/tokenization, or explicitly rename and type the value as `charDelta` everywhere; (2) expose generated JSON/HTML reports from the supported training status/CLI surface with stable retrievable references; (3) remove unsupported placeholder sections rather than presenting them as implemented; and (4) capture and manually inspect one real optimizer report. Provider/render experimentation belongs with narrowed #14875 or a future non-MVP proposal. Remove this from Project 15; #16935–#16945 cover current LifeOps verification scope.

**Comment by kenjohnscreates** [comment]
Staleness validation — recommend closing or re-scoping. The narrowed acceptance criteria (2026-07-23 triage) target code that no longer exists on `develop`:

- `plugins/plugin-training` (69 files, including `src/core/optimization-report.ts` from #15100) was deleted in `dc03207dc6` ("delete a bunch of stuff", 2026-08-01) and remains absent on current `develop`.
- The root `bun run train` script referenced by the remaining acceptance bar no longer exists in `package.json`.

Verified on current `develop` head today. Every remaining acceptance item (real token metric or `charDelta` rename, report discoverability from the training status/CLI surface, placeholder-section removal, captured optimizer report) has no surviving implementation surface. If optimizer reporting returns in a future training stack, a fresh issue scoped to that architecture seems cleaner than keeping this one open — deferring the disposition to maintainers.

AI provider/model: anthropic / claude-fable-5
Client / agent tooling: claude-code
Contribution skill revision: elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza
Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
Attribution status: self-reported
— [kenjohnscreates]

**Comment by lalalune** [comment]
Closing as obsolete rather than rebuilding a deleted subsystem. PR #15100 implemented the requested optimizer report persistence, after which the repository deliberately removed `plugin-training` and the root `train` command. There is no current owning runtime or supported user path on which the remaining historical acceptance items can operate. Any future training product should start from its current architecture and evidence contract, not revive this stale issue.

AI provider/model: OpenAI / gpt-5.6-sol
Client / agent tooling: Codex desktop
Contribution skill revision: elizaOS/eliza@b9dec8f6054b9d563be68fbb8421bb95b18cd977:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [codex]
